queue/ready-queue.js

```diff
@@ -159,6 +196,63 @@ module.exports = function (RED) {
 		// call of sqlite opening the db file
 		node.on('input',initialState) ;
 
+		/**
+		 * Get the file size of the provided filename in MB
+		 * useful to check whether the DB went over a given size
+		 */
+		function getFilesizeInMegaBytes(filename) {
+			var stats = fs.statSync(filename);

```

Comment:
Should never call synchronous methods as they will block the node event thread.  

You will need to change to a non-blocking call on stat here and return a promise instead

Reply:
I understand your concerns about blocking the node event thread. I share them. I thought in that instance that was acceptable for the following reasons:

- At the Node initialization level [above](https://github.com/damoclark/node-red-contrib-msg-queue/pull/10/files#diff-fddf465c13f4805a01f26ebdcb1b9d76R121), we need a synchronized call to check the size. If the max filesize was provided and is below the current file size, then we want to display a status update and stop the flow. [Node-Red Doc](https://nodered.org/docs/user-guide/writing-functions#handling-errors) `If the function encounters an error that should halt the current flow, it should return nothing.`. If we were to implement a Promise-based function, then the flow would carry on and not stop which is not desirable.

- Because of the first point, we could then use two implementations of the two functions `getFilesizeInMegaBytes` and `isAboveDBFileSizeLimit`: one would be synchronous and the other one asynchronous. Looking at it, I thought that was a bit overkilled because, after the node initializes, this function would always be called in a `setTimeout`. Therefore it is already outside of the main even loop and would be handle by another CPU core if you have one.

- I also looked how much time was required for this call to perform on really small hardware for a really big file: 35ms. 

In conclusion, because this call is fast, is executed only every 10 sec, in a different even loop, I thought that was acceptable. Please let me know what you think? Do you think I should still implement two other async methods for `getFilesizeInMegaBytes` and `isAboveDBFileSizeLimit`?

Thank you!

Reply:
I got a bit more interested in these 35ms for the execution of the `fs.statSync`. I wanted to see how that compared to the execution of other synchronous code we have in the project.

So I bench-marked the synchronous `processStatus` method we have in this module on the same small IoT hardware with a simple:

```
  var date1 = new Date().getTime()
  processStatus(msg) ;
  var date2 = new Date().getTime()
  node.log(`milliseconds processStatus: ${date2 - date1}`)
```

And I did the same for `getFilesizeInMegaBytes` but this time with a DB file, not with a random large file.

Here is the result that I had consistently:
```
milliseconds getFilesizeInMegaBytes: 12
milliseconds processStatus: 60
```

Reply:
> I understand your concerns about blocking the node event thread. I share them. I thought in that instance that was acceptable for the following reasons:
> 
> At the Node initialization level [above](https://github.com/damoclark/node-red-contrib-msg-queue/pull/10/files#diff-fddf465c13f4805a01f26ebdcb1b9d76R121), we need a synchronized call to check the size. If the max filesize was provided and is below the current file size, then we want to display a status update and stop the flow. [Node-Red Doc](https://nodered.org/docs/user-guide/writing-functions#handling-errors) `If the function encounters an error that should halt the current flow, it should return nothing.`. If we were to implement a Promise-based function, then the flow would carry on and not stop which is not desirable.

You are quoting from the [manual](https://nodered.org/docs/user-guide/writing-functions) for the built-in `Function` node, rather than the manual for writing your own node.  In the context of that manual page, halting the flow relates to the flow of the current message being processed by the function, rather than halting the flow tab in its entirety. 

There is no harm to asynchronous initialisation of a custom node, except that once it is registered, it will start receiving messages as explained by dceejay in his response to a similar query on the [discourse page](https://discourse.nodered.org/t/custom-node-waiting-for-instance-to-be-ready/14400).

There is little to be gained by making synchronous IO calls for the node initialisation.

D.

Reply:
Thanks for these links @damoclark and for the time you put into this, really appreciate it!

Regarding this part here:

> There is little to be gained by making synchronous IO calls for the node initialisation.

In my opinion it was needed, I will try to explain why. When the Node starts I wanted to do a primarily check on the file size. I looked in the project if there was something similar. And there was: the check on the filename which is done [here](https://github.com/damoclark/node-red-contrib-msg-queue/pull/10/files#diff-fddf465c13f4805a01f26ebdcb1b9d76R110-R112)

I believe these two belong to the same family: validations. Usually when you do validations, you error and exit if something does not pass the validation. Looking at the filename validation, it was weird to me that we were doing this test, logging an error when no filename was provided but still the flow would carry on as if nothing happened. 
Way later in the [initialization process](https://github.com/damoclark/node-red-contrib-msg-queue/pull/10/files#diff-fddf465c13f4805a01f26ebdcb1b9d76R407-R414), we are trying to open this filename that does not exist and catch the error. I think we could have returned earlier, but it is working so I left as is.

For the file-size validation, I wanted to check the size and if it was too big show an error and stop the flow. If I do it asynchronously, then the rest of the initialization would carry on (same way as it carries on for the file-name validation). This means that if the flow is started with a connection status "down", it will start to add more messages to the file which already exceeds the max file size. In my opinion, this was not desirable.

Reply:
Updated: links in the above comment.

Reply:
Hi again @olivierpichon ,

> Thanks for these links @damoclark and for the time you put into this, really appreciate it!
> 

Likewise, I appreciate the time you have taken to prepare a PR to improve it.  My apologies for the delay in responding.  This work is a hobby - I have to pay the bills first.

> Regarding this part here:
> 
> > There is little to be gained by making synchronous IO calls for the node initialisation.
> 
> In my opinion it was needed, I will try to explain why. When the Node starts I wanted to do a primarily check on the file size. 
> ...
>  Usually when you do validations, you error and exit if something does not pass the validation. Looking at the filename validation, it was weird to me that we were doing this test, logging an error when no filename was provided but still the flow would carry on as if nothing happened.
> ...
> For the file-size validation, I wanted to check the size and if it was too big show an error and stop the flow. 

This is where I think there is confusion.  When you say 'stop the flow', are you meaning that you wish to disable an entire flow tab from processing messages?  If so, to my knowledge, the NR API does not allow a node at initialisation to do this.  Things may have changed since I wrote this node, but I did a quick search and couldn't anything in the documentation to do this.  In your [earlier response](https://github.com/damoclark/node-red-contrib-msg-queue/pull/10#discussion_r466875616) to my query about calling blocking functions, you referred to the [Node-Red Doc](https://nodered.org/docs/user-guide/writing-functions#handling-errors) and said the following:

> If the function encounters an error that should halt the current flow, it should return nothing.. If we were to implement a Promise-based function, then the flow would carry on and not stop which is not desirable.

That page of the Node-Red Doc and the quote that you reference pertains to users of NR writing javascript inside of a Node-Red Function node.  And when it mentions stopping the flow, it means that the message it is currently processing stops (is discarded), rather than disabling the entire flow tab.  It is part of the NR 'user guide', not the 'creating nodes' guide that is relevant to this work.

If I have misunderstood or am unaware of new features in NR that allow you to disable the entire flow as I interpret your intention with the validation step, please let me know.  Otherwise, as I previously stated, there is little to be gained in calling blocking functions.  

>  I also looked how much time was required for this call to perform on really small hardware for a really big file: 35ms.
>
> In conclusion, because this call is fast, is executed only every 10 sec, in a different even loop, I thought that was acceptable. Please let me know what you think? Do you think I should still implement two other async methods for getFilesizeInMegaBytes and isAboveDBFileSizeLimit?

I am very wary of adding any blocking calls no matter how much benchmarking you undertake as its just bad practice.  There is only one event loop in NR that I am aware of, and for that 35ms (which is actually a considerable amount of CPU time), Node Red can't do any work.  From memory, your implementation does this every 10 seconds, not just in the initialisation of the node.

In your description of the changes, you mention:

> If for any reasons, the limit is reached, we forward the messages to the next nodes but no longer store them in the DB.

One of the promises of the node is that it will process queued messages in the same order that they arrived.  This would contravene the current description of the node's characteristics.  Preferably, I would want the node to dispatch a message from the queue, as messages are added to the queue.  I do recognise that if new messages are larger than old ones, the queue on disk will continue to grow.  It's a tricky balance.  Happy to discuss the behaviour when disk is full further.

In summary, I would like to commit your contribution to the node as your proposed improvement is a good idea.  But you will need to address the issues I have raised - especially the inclusion of blocking IO calls that at this stage, I am unconvinced are necessary.

Regards,
Damien.

Reply:
> That page of the Node-Red Doc and the quote that you reference pertains to users of NR writing javascript inside of a Node-Red Function node. And when it mentions stopping the flow, it means that the message it is currently processing stops (is discarded), rather than disabling the entire flow tab. It is part of the NR 'user guide', not the 'creating nodes' guide that is relevant to this work.

You're right, I did quote the wrong documentation page, sorry about that. I do believe the principle remains though. There need to be a set of pre-requisites at the node startup for it to behave correctly. Before the changes introduced in this very PR, this pre-requisite was: "the db file needs to be not empty and also needs to exist". This is outside of this PR but I believe that the last part of the pre-requisite (the file needs to exist) could be removed if `node-persistent-queue` created the DB file when it does not exist [here](https://github.com/damoclark/node-persistent-queue/blob/master/index.js#L224-L232)

Getting back to this PR, it introduces limit with the filesize. So a new pre-requisite for the node to start correctly is that the DB file is no bigger than the limit (if limit set).

Let's see what's happening with the first pre-requisite, *before the changes introduced in this PR*. I put an empty file name in the configuration, deploy and look at the debug output:
<img width="1285" alt="Screen Shot 2020-08-20 at 12 00 18" src="https://user-images.githubusercontent.com/3260198/90756622-e623a680-e2dc-11ea-8321-29a8f4ab5098.png">
As you can see, there is no output in the debug pane, the flow is actually stopped. We do not forward messages to the next node. That's what I mean when I said "stopped". However, please note the status of the node Queue, it says "Forwarding" in green.

The reason for it not to be forwarding messages to the next node when the DB file is empty is that [we register the processing of node input](https://github.com/damoclark/node-red-contrib-msg-queue/blob/master/queue/ready-queue.js#L328-L342) in `queue.on('open')`

When the filename is empty, `queue.on('open')` fails [here](https://github.com/damoclark/node-red-contrib-msg-queue/blob/master/queue/ready-queue.js#L299-L306)
And so the queue's `open` event is never fired [there](https://github.com/damoclark/node-red-contrib-msg-queue/blob/master/queue/ready-queue.js#L309)

What could be the alternative to this:

1. One alternative could be to [catch](https://github.com/damoclark/node-red-contrib-msg-queue/blob/master/queue/ready-queue.js#L303) when we can't open the DB and still forward messages to the next node in that case. I don't find that it is a good approach because for the end user it would look like the flow and the specific node works as expected while it is obviously not the case: it won't be able to persist anything in DB in case it needs to.
2. The other alternative is, as we stop processing messages anyway, (or more so never processed them), to return early: as early as the test is done [here](https://github.com/damoclark/node-red-contrib-msg-queue/blob/master/queue/ready-queue.js#L87-L89) and show a status message error.

As you know, I favoured the second approach in this PR [to check the file size](https://github.com/damoclark/node-red-contrib-msg-queue/pull/10/files#diff-fddf465c13f4805a01f26ebdcb1b9d76R121-R128). I chose to return, ie giving up on processing the messages as early as I can and show a message to the end user, that they can then address. As discussed, this also prevents from the file getting bigger at every launch if the connectivity is stopped and the file is already bigger than the max allowed file size. The drawback is that this test *needs to be synchronous*.

> I am very wary of adding any blocking calls no matter how much benchmarking you undertake as its just bad practice.

I respectfully disagree with this. Practices, good or bad are tools. If we discard something just because of "Bad Practice" without considering the conditions in which something is running, this starts to become more like a religion. The benchmarks aimed just to look at it from an angle that we could count, not on a belief.

After running and testing the node multiple times, I still believe the approach in the PR is the tradeoff that addresses most of the important points. I realise I may not have been successful in convincing you which is 100% my failure. And so I understand you may want to close this PR then.

In any cases, thanks again for the time you took reviewing this and also for authoring these two modules!

Reply:
Hi again @damoclark ,

Thanks a lot for your answer. So I looked a bit more into your comment and tried to do more tests:

> This line
>
> this.db = new sqlite3.Database(this.dbPath, sqlite3.OPEN_READWRITE | sqlite3.OPEN_CREATE, err => {
>
> says to open this.dbPath file read/write, and if it doesn't exist, create it sqlite3.OPEN_CREATE. No error occurs if the file doesn't exist - sqlite will just create it. An error will occur if it is unable to open an existing file or unable to create the file, typically due to permissions or an invalid path. In such cases, the node.error method is called which generates an error message to the console and the debug pane of the UI alerting the user that the node is not working.

You are absolutely right. It will create the file if it does not exist. It will also raise an error if unable to open an existing file or unable to create one, all good 👍 

> The user is notified that the node is misconfigured, and as you have explained, discards messages.

> >    What could be the alternative to this:
> >
> >    2. The other alternative is, as we stop processing messages anyway, (or more so never processed them), to return early: as early as the test is done [here](https://github.com/damoclark/node-red-contrib-msg-queue/blob/master/queue/ready-queue.js#L87-L89) and show a status message error.

Around this point above here, so I originally thought that we indeed discard messages after we attempt to open the DB:

```javascript
queue.open()
.then(function() {
  node.log("Opened " + node.sqlite + " successfully.") ;
})
.catch(function(err) {
  node.error("Queue failed to open " + node.sqlite, err);
  // @todo Check does this handle sqlite open error condition accordingly to node-red framework
}) ;
```

Basically my reasoning was that we would enter the `catch` method when giving an empty file, same way we enter the catch method when we give a file that can't be opened.

Then:
```javascript
queue.on('open',function() {
...
```
would never be executed and the messages would then be discarded.

But as I tested it further, I realised that it is not what is happening. When no file is provided, `node-persistent-queue` will use an in-memory store and will process messages as if everything was behaving correctly, we enter the `then` function of the promise here:

```javascript
queue.open()
.then(function() {
  node.log("Opened " + node.sqlite + " successfully.") ;
})
```

And the log file shows:
```
24 Aug 12:59:54 - [info] Starting flows
24 Aug 13:00:00 - [error] [queue:Queue] No filename specified for the queue SQLite DB 
24 Aug 13:00:00 - [info] Started flows
24 Aug 13:00:00 - [info] [mqtt-broker:enspara-staging] Connection failed to broker: mqtts://a2rwmfnnwq2v8b-ats.iot.us-east-1.amazonaws.com:8883
24 Aug 13:00:00 - [info] [queue:Queue] Opened  successfully
```

On line 2 of the above log, we have the `node.error` due to the very early validation:
```javascript
// Generate error if filename to sqlite db not provided
if(node.sqlite === undefined || node.sqlite == '') {
	node.error("No filename specified for the queue SQLite DB") ;
}
```

But on the last line of the log, we have `[info] [queue:Queue] Opened  successfully` (with an empty DB file) coming from: `node.log("Opened " + node.sqlite + " successfully.") ;`

I was relying on the fact that we never get any further processing messages when the filename is not provided. So these tests help me understanding that there was a bug then in my previous implementation: if no filename was provided and a max file size was provided: the flow is not interrupted and it would check the file size of an empty file which, of course, does not go well.

So I committed a change to this PR where we are a bit more defensive, checking if the file name is provided when checking the file size. 

I also made the change for the file size check to be asynchronous as requested: this takes place on the promise chain, right after we attempt to open the DB.

Let me know what you think.

Thanks again!